Event System Framework

liamjwarren opened the pull request:
The event system framework is now put in place. I still need to connect it with the callback functions and the input polling to ensure it functions as intended. I also want to make a function to set all callbacks that are currently done in Window.cpp to be a function in the Callbacks.h file. I will refine further in the coming days but wanted to post to get some comments to date and see if there are any glaring issues that I need to address.

liamjwarren committed:
Restructuring

-Renamed files to ensure clarity.
-Created additional files that will be implemented to refine the initial event system.
-Commit restructuring changes prior to implementing further changes.

liamjwarren committed:
Restructuring

-Renamed files to ensure clarity.
-Created additional files that will be implemented to refine the initial event system.
-Commit restructuring changes prior to implementing further changes.

liamjwarren committed:
Restructuring

-Renamed files to ensure clarity.
-Created additional files that will be implemented to refine the initial event system.
-Commit restructuring changes prior to implementing further changes.

liamjwarren committed:
Restructuring

-Renamed files to ensure clarity.
-Created additional files that will be implemented to refine the initial event system.
-Commit restructuring changes prior to implementing further changes.

liamjwarren committed:
Restructuring

-Renamed files to ensure clarity.
-Created additional files that will be implemented to refine the initial event system.
-Commit restructuring changes prior to implementing further changes.

liamjwarren committed:
Restructuring

-Renamed files to ensure clarity.
-Created additional files that will be implemented to refine the initial event system.
-Commit restructuring changes prior to implementing further changes.

liamjwarren committed:
Commit prior to rebasing on master

liamjwarren committed:
Commit prior to rebasing on master

liamjwarren committed:
Commit prior to rebasing on master

liamjwarren committed:
Rebased and changed includes in Window.cpp

liamjwarren committed:
Rebased and changed includes in Window.cpp

liamjwarren committed:
Merge branch 'refine-initial-event-system' of https://github.com/Burke-Daniel/Aquarius into refine-initial-event-system

liamjwarren committed:
Revert "Merge branch 'refine-initial-event-system' of https://github.com/Burke-Daniel/Aquarius into refine-initial-event-system"

This reverts commit b86cdc580f372fa89faf564774a955949d180002, reversing
changes made to 24fdf5f5d37083eae6d158cfb8518e4efe843a8a.

liamjwarren committed:
Revert "Merge branch 'refine-initial-event-system' of https://github.com/Burke-Daniel/Aquarius into refine-initial-event-system"

This reverts commit b86cdc580f372fa89faf564774a955949d180002, reversing
changes made to 24fdf5f5d37083eae6d158cfb8518e4efe843a8a.

liamjwarren committed:
Revert "Merge branch 'refine-initial-event-system' of https://github.com/Burke-Daniel/Aquarius into refine-initial-event-system"

This reverts commit b86cdc580f372fa89faf564774a955949d180002, reversing
changes made to 24fdf5f5d37083eae6d158cfb8518e4efe843a8a.

liamjwarren committed:
Commit work prior to switching to implementing Observer based system.

liamjwarren committed:
Commit prior to refactoring subscriber format

liamjwarren committed:
Rebased on Master

liamjwarren committed:
Before pulling origin

liamjwarren committed:
Merge branch 'refine-initial-event-system' of https://github.com/Burke-Daniel/Aquarius into refine-initial-event-system

liamjwarren committed:
Resolving issues to run

liamjwarren committed:
Merge branch 'refine-initial-event-system' of https://github.com/Burke-Daniel/Aquarius into refine-initial-event-system

liamjwarren committed:
Reverted to modifications

liamjwarren committed:
Remove unnecessary log includes for now

colton-smith commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,135 @@
+        static void cursorPositionCallback(GLFWwindow* window, double xPosition, double yPosition)
+        {
+            // TODO: Discuss if we want to only keep track of cursor position when the cursor is on
+            //  the Application window, if so can combine this callback method and the one below.
```
Definitely dont care if the cursor is outside the window I'd say

Burke-Daniel replied:
^^

colton-smith commented on Aquarius/Src/Aquarius/Events/Event.h:
```diff
@@ -0,0 +1,31 @@
+            return type;
+        }
+
+        virtual bool handledStatus()
```
Not sure this needs to be virtual, should be the same function definition for all sub classes

colton-smith replied:
also isHandled might be a little clearer of a name

colton-smith commented on Aquarius/Src/Aquarius/Core/Window.cpp:
```diff
@@ -1,9 +1,8 @@
 #include "Window.h"
+#include "Log.h"
+#include <assert.h>
+#include "Aquarius/Events/Callbacks.h"
```
Should be:
-  Aquarius/Core/Log.h
-  assert.h is incorrectly placed and shouldn't be needed anyway

colton-smith commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,135 @@
+
+        static void mouseButtonCallback(GLFWwindow* window, int button, int action, int modifications)
+        {
+            switch (action)
+            {
+                //TODO: Get it so mouse button name is printed rather then number, this is not critical but would be
+                // nice to have.
+                case GLFW_PRESS:
+                    AQ_INFO("mouse button %v is pressed.", button);
+                    break;
+
+                case GLFW_RELEASE:
+                    AQ_INFO("mouse button %v has been released.", button);
+                    break;
+
+                default:
+                    break;
+            }
```
I guess these callbacks will need to emit an event of appropriate type right? Either directly use a singleton instance of EventManager or pass the event to Application where it can then call EventManager->Notify

liamjwarren replied:
Yes, something along that line.

Burke-Daniel replied:
If EventManager is a member of Application, we can use the static Application get() method to accomplish that I would say. Just one way of going about it

colton-smith commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,135 @@
+        static void windowCloseCallback(GLFWwindow* window)
+        {
+            //TODO Implement
+        }
```
I'd say we can cut these down to what is really necessary for us

colton-smith commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,42 @@
+    private:
+        using functionType = std::function<void(const Event<T>&)>;
+        std::map<T, std::vector<functionType>> subscribers;
+
```
When you had this after public, the public functions likely had no idea what functionType was. One option is to just move `using functionType ... ` above the public stuff

colton-smith commented on Aquarius/Src/Aquarius/Events/Event.h:
```diff
@@ -0,0 +1,31 @@
+
+namespace Aquarius {
+
+    template <typename T>
```
Might need someone to explain to me why this class needs to be templated

colton-smith replied:
Oh, i think its because the event types are not all in the same enum

liamjwarren replied:
Correct

colton-smith commented on Aquarius/Src/Aquarius/Events/MouseEvent.h:
```diff
@@ -0,0 +1,45 @@
+        MouseButtonPressedEvent, MouseButtonReleasedEvent, MouseMovedEvent, MouseScrolledEvent, MouseInWindowEvent
+    };
+
+    class MouseButtonPressed : public Event<MouseEvents>
```
Wondering if MouseButtonPressedEvent (suffix Event) might be a better way to name these events

Burke-Daniel commented:
Haven't actually reviewed this fully yet, but saw a chance to use macros to reduce code repetition when defining the derived event classes for each event type. See this [branch](https://github.com/Burke-Daniel/Aquarius/tree/refine-initial-event-system-review) for what I did, and if you think it's reasonable then you can just cherry-pick the commit, if not then disregard lol

colton-smith commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,42 @@
+    {
+    private:
+        using functionType = std::function<void(const Event<T>&)>;
+        std::map<T, std::vector<functionType>> subscribers;
```
We should also consider using unordered_map, it's much more performant and I **think** will serve the same purpose

colton-smith commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,42 @@
+        {
+            // Check to see if at end of subscribers
+            if((subscribers.find(event.getEventType())) == subscribers.end())
+                return;
```
This checks if there are any subscribers of a given event type, end of subscribers was a little misleading.

colton-smith commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,135 @@
+        }
+    };
+
+    class MouseEvent
```
I think we should update these names to be MouseCallback or something. I do like defining them here and using them in the window class to register the callbacks

Burke-Daniel replied:
Yeah and on that note, do each of these need to be within a class or can they just be free functions?

colton-smith reviewed (changes requested):
Very sick my guy! I think I understand how it will work, just need to plug it in in a few places and we're g2g with events!

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,135 @@
+            // TODO: Discuss if we want to only keep track of cursor position when the cursor is on
+            //  the Application window, if so can combine this callback method and the one below.
+            AQ_INFO("x position: %v", xPosition);
+            AQ_INFO("y position: %v", yPosition);
```
Do you wanna drop the log level of this to TRACE? These will definitely just be used for debugging as they pollute the logs haha

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,135 @@
+        static void mouseScrollCallback(GLFWwindow* window, double xOffSet, double yOffset)
+        {
+            AQ_INFO("x offset: %v", xOffSet);
+            AQ_INFO("y offset: %v", yOffset);
```
Could drop these to TRACE as well

Burke-Daniel replied:
Also I think any logging in here we will probably want to be CORE

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,42 @@
+            for(auto&& subscriber : subscribers.at(event.getEventType()))
+            {
+                // Notify of event only if the event has not already been handled
+                if(event.handledStatus() == 0)
```
Make comparison to bool instead of int

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,42 @@
+    public:
+        void subscribe(T Type, const functionType& function)
+        {
+            subscribers[Type].push_back(function);
```
For subscribers where the event will change the state of the object, we will wanna make sure to use std::bind, and bind member functions and a respective instantiated object here

liamjwarren replied:
Do you mean to use std::bind here or when we invoke it? I think you mean when we invoke it based on reading this [Article](https://en.cppreference.com/w/cpp/utility/functional/bind )

Burke-Daniel replied:
When we invoke it yeah, this was just me thinking out loud haha

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/KeyboardEvent.h:
```diff
@@ -0,0 +1,33 @@
+        KeyPressedEvent, KeyRepeatedEvent, KeyReleasedEvent,
+    };
+
+    class KeyPressed : public Event<KeyboardEvents>
+    {
+    public:
+        KeyPressed() : Event<KeyboardEvents>(KeyboardEvents::KeyPressedEvent) {};
+    };
+
+    class KeyRepeated : public Event<KeyboardEvents>
+    {
+    public:
+        KeyRepeated() : Event<KeyboardEvents>(KeyboardEvents::KeyRepeatedEvent) {};
+    };
+
+    class KeyReleased : public Event<KeyboardEvents>
+    {
+    public:
+        KeyReleased() : Event<KeyboardEvents>(KeyboardEvents::KeyReleasedEvent) {};
+    };
```
As discussed, will have to add data about the type of key pressed

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/MouseEvent.h:
```diff
@@ -0,0 +1,45 @@
+        MouseButtonPressedEvent, MouseButtonReleasedEvent, MouseMovedEvent, MouseScrolledEvent, MouseInWindowEvent
+    };
+
+    class MouseButtonPressed : public Event<MouseEvents>
+    {
+    public:
+        MouseButtonPressed() : Event<MouseEvents>(MouseEvents::MouseButtonPressedEvent) {};
+    };
+
+    class MouseButtonReleased : public Event<MouseEvents>
+    {
+    public:
+        MouseButtonReleased() : Event<MouseEvents>(MouseEvents::MouseButtonReleasedEvent) {};
+    };
+
+    class MouseMoved : public Event<MouseEvents>
+    {
+    public:
+        MouseMoved() : Event<MouseEvents>(MouseEvents::MouseMovedEvent) {};
+    };
+
+    class MouseScrolled : public Event<MouseEvents>
+    {
+    public:
+        MouseScrolled() : Event<MouseEvents>(MouseEvents::MouseScrolledEvent) {};
+    };
+
+    class MouseInWindow : public Event<MouseEvents>
+    {
+    public:
+        MouseInWindow() : Event<MouseEvents>(MouseEvents::MouseInWindowEvent) {};
+    };
```
Same as above, will have to add info about what button is pressed

Burke-Daniel reviewed (changes requested):
A lot of these changes we already discussed today, so once those are changed I'd like to have another look. Also, seems like the commit history got a bit messy there. No big deal but if it happens in the future you could do an interactive rebase and tidy it up. This event system is super nice!

liamjwarren committed:
Changes made based on reviews

liamjwarren commented:
Thanks for the review guys! Just pushed a commit with the changes we discussed. I added in the following parameters for events: code which is used for key or mouse codes and xAxis & yAxis which is used for mouse scrolling and position. I also added in some getters for these in the events. I had to leave the static in front of the callback function definitions as when I removed it I was receiving a linking error. Let me know what you guys think and if you have any questions!

liamjwarren commented:
Once I get approvals I am planning on merging this branch and then making a new one to try and implement it as we discussed FYI.

colton-smith commented on Aquarius/Src/Aquarius/Core/Window.cpp:
```diff
@@ -64,11 +63,11 @@ namespace Aquarius {
+        glfwSetCursorPosCallback(m_Window, mouseCursorPositionCallback);
+        glfwSetCursorEnterCallback(m_Window, mouseCursorInWindowCallback);
+        glfwSetMouseButtonCallback(m_Window, mouseButtonCallback);
+        glfwSetScrollCallback(m_Window, mouseScrollCallback);
```
Very nice! Might still want to consider placing those guys in a callbacks:: namespace or similar though!

colton-smith commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,73 @@
+
+                default:
+                    break;
+            }
```
The callbacks have to emit events

colton-smith commented on Aquarius/Src/Aquarius/Events/Event.h:
```diff
@@ -0,0 +1,54 @@
+        T m_Type;
+        bool m_Handled = false;
+
+        // The below are parameters that are used to provide more information about the event.
+        // If the parameter is not applicable to the event it will be set to 0.
+        // Code is used for Mouse codes & Key codes.
+        int m_Code;
+        double m_xAxis;
+        double m_yAxis;
```
These members belong in the derived event classes, not here

Burke-Daniel replied:
^^ This will prevent the members from never being necessary, as they will only be present in events which contain them as information

colton-smith commented on Aquarius/Src/Aquarius/Events/KeyboardEvent.h:
```diff
@@ -0,0 +1,33 @@
+        KeyPressedEvent() : Event<KeyboardEvents>(KeyboardEvents::KeyPressedEvent, m_Code, 0,0) {};
+    };
+
+    class KeyRepeatedEvent : public Event<KeyboardEvents>
+    {
+    public:
+        KeyRepeatedEvent() : Event<KeyboardEvents>(KeyboardEvents::KeyRepeatedEvent, m_Code, 0, 0) {};
+    };
```
Give these classes the private data members and getters/setters

colton-smith commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,73 @@
+
+
+namespace Aquarius {
+
```
Probably a good idea: Callbacks:: namespace

colton-smith reviewed (changes requested):
Will need to move data members out of the base event class and into the event sub classes, and add some getter/setter methods to the sub classes as well.

Also, consider adding a namespace to the callbacks and also adding event creation and emitting to the appropriate callbacks.

colton-smith commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,42 @@
+
+#include "Event.h"
+
+#include <GLFW/glfw3.h>
+#include <unordered_map>
+#include <vector>
```
Cant get the engine to build without including <functional>

colton-smith commented on Aquarius/Src/Aquarius/Events/KeyboardEvent.h:
```diff
@@ -0,0 +1,33 @@
+    class KeyPressedEvent : public Event<KeyboardEvents>
+    {
+    public:
+        KeyPressedEvent() : Event<KeyboardEvents>(KeyboardEvents::KeyPressedEvent, m_Code, 0,0) {};
```
This is kinda crazy, you are initializing a class with its own member? It seems cyclical in nature and very not sure how it works>

Burke-Daniel replied:
Yeah I feel like this is very dangerous, Better to add the member variables in to these derived classes so that the callbacks can pass them in to these constructors when they create the appropriate event objects

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/Event.h:
```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <GLFW/glfw3.h>
```
I don't think this needs to be included here

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,42 @@
+
+#include "Event.h"
+
+#include <GLFW/glfw3.h>
```
Ditto here

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,42 @@
+            for(auto&& subscriber : subscribers.at(event.getEventType()))
+            {
+                // Notify of event only if the event has not already been handled
+                if(!event.isHandled())
+                {
+                    subscriber(event);
+                }
```
So to me, this looks like a client can set the handled member variable to true, and then nothing else occurs. If the event is handled, does it make sense to just break out of this for loop? This will just loop through the rest of the vector and do nothing if the event has already been handled by one of the other subscribers

colton-smith replied:
@Burke-Daniel I think we should break when it gets handled for sure. However, some events may need information from an event but not actually consume the event, therefore we shouldnt break after line 33. Maybe we add a check after line 34, if (handled) {break;}

Burke-Daniel reviewed (changes requested):
Couple more small things then I'd say you can merge after another look and go ahead and add the event creation to the callbacks, add event handlers to the application class, etc.

liamjwarren committed:
Switch from using templates to inheritance for event system

liamjwarren commented:
Alright, this is what I have gotten done to date. I still need to troubleshoot the event handler portion, but I now have the window and key callback’s now creating an event and logging this correctly. 

I will troubleshoot the event handler tomorrow; I need to find the best way to tie it into the application and make sure it actually works as I tried a few different ways tonight and was getting seg faults. Once I have this done, I think everything else should fall into place in terms of subscriptions and notifications. I am thinking it may be good for testing purposes to create some unit tests or assertions for the subscriptions and notification functions are working as intended.

Let me know if you see any glaring issues or if you have any questions. I will try and get another updated commit up tomorrow as well with tying in the event handler.

colton-smith commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,81 @@
+
+        static void mouseScrollCallback(GLFWwindow *window, double xOffSet, double yOffset)
+        {
+            MouseScrolledEvent event(xOffSet, yOffset);
```
OffSet and offset

colton-smith commented on Aquarius/Src/Aquarius/Events/Event.h:
```diff
@@ -0,0 +1,31 @@
+
+#define EVENT_TYPE(type) static eventType getStaticType() { return eventType::type; }\
+                                                            virtual eventType getEventType() const override { return getStaticType(); }\
+                                                            virtual const char* getName() const override { return #type; }
```
As much as this is useful, it may be okay to just write these all out. Also, we can probably do std::string and not const char pointer

colton-smith commented on Aquarius/Src/Aquarius/Events/Event.h:
```diff
@@ -0,0 +1,31 @@
+        bool isHandled()
+        {
+            return m_Handled;
+        };
```
m_Handled should be private otherwise the getter doesn't really make sense

colton-smith commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,55 @@
+
+        EventHandler(Event& event) : m_event(event)
+        {
+            logCreation();
```
Not too sure we need to keep log messages in constructors in release / master

Burke-Daniel replied:
Can probably wrap in `#ifdef AQ_DEBUG` or something

colton-smith commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,55 @@
+            logCreation();
+        }
+
+        void logCreation()
+        {
```
I dont really mind this method tho, its essentially like a to_string method

colton-smith commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,55 @@
+        virtual ~EventHandler() = default;
+
+    private:
+        std::unordered_map<eventType, std::vector<functionType>> subscribers;
```
should be prefixed m_

colton-smith commented on Aquarius/Src/Aquarius/Events/KeyboardEvent.h:
```diff
@@ -0,0 +1,67 @@
+    {
+    public:
+
+        int getCode()
```
This method should be const, and we should be using Input::KeyCode enum for the keycode member

Generally, any method that won't modify the class should be marked  const

colton-smith commented on Aquarius/Src/Aquarius/Events/KeyboardEvent.h:
```diff
@@ -0,0 +1,67 @@
+
+        EVENT_TYPE(KeyPressedEvent)
+
+        void logEvent()
```
Looks as tho every subclass of event has a logEvent class, maybe it should be declared virtual in the base event class?

Burke-Daniel replied:
^^ I agree, probably don't make it purely virtual though so that the derived class doesn't have to override it (could just be a virtual function with an empty body)

colton-smith commented on Aquarius/Src/Aquarius/Events/KeyboardEvent.h:
```diff
@@ -0,0 +1,67 @@
+        }
+
+    protected:
+        int m_repeatCount;
+    };
```
There is no way for an external caller to access this member because it is protected. I doubt this will be subclassed so repeat count could be private (with a getter to provide access).

colton-smith commented on Aquarius/Src/Aquarius/Events/MouseEvent.h:
```diff
@@ -0,0 +1,135 @@
+    {
+    public:
+        MouseMovedEvent(float xCoordinate, float yCoordinate)
+                : m_XCoordinate(xCoordinate), m_YCoordinate(yCoordinate) {}
```
I think just m_x and m_y would suffice? Or x_pos? x_coord? xCoordinate seems a little bit verbose.

For me to get the x/y of a mouse, I would need to call:
```
mouseMovedEvent.getXCoordinate()
mouseMovedEvent.getYCoordinate()
```

colton-smith replied:
I would also add a getter that returns a glm::vec2 or std::pair of the (x,y) coordinates

colton-smith commented on Aquarius/Src/Aquarius/Events/MouseEvent.h:
```diff
@@ -0,0 +1,135 @@
+            AQ_CORE_TRACE("MouseMovedEvent: %v, %v", m_XCoordinate, m_YCoordinate);
+        }
+
+        float getXCoordinate()
+        {
+            return m_XCoordinate;
+        }
+
+        float getYCoordinate()
+        {
+            return m_YCoordinate;
+        }
```
const

colton-smith commented on Aquarius/Src/Aquarius/Events/MouseEvent.h:
```diff
@@ -0,0 +1,135 @@
+        float m_YCoordinate;
+    };
+
+    class MouseScrolledEvent : public Event
+    {
+    public:
+        MouseScrolledEvent(float xCoordinate, float yCoordinate)
+                : m_XCoordinate(xCoordinate), m_YCoordinate(yCoordinate) {}
+
+        EVENT_TYPE(MouseScrolledEvent)
+
+        void logEvent()
+        {
+            AQ_CORE_TRACE("MouseScrolledEvent: %v, %v", m_XCoordinate, m_YCoordinate);
+        }
+
+        float getXCoordinate()
+        {
+            return m_XCoordinate;
+        }
+
+        float getYCoordinate()
+        {
+            return m_YCoordinate;
+        }
+
+    private:
+        float m_XCoordinate;
+        float m_YCoordinate;
+    };
```
Maybe make the member names better reflect a mouse scroll, this is a copy+paste of the MouseMovedEvent

colton-smith commented on Aquarius/Src/Aquarius/Events/MouseEvent.h:
```diff
@@ -0,0 +1,135 @@
+        }
+
+    protected:
+        int m_buttonCode;
```
Use the enum in Input.h instead of ints again

colton-smith commented on Aquarius/Src/Aquarius/Events/WindowEvent.h:
```diff
@@ -0,0 +1,17 @@
+
+
+namespace Aquarius {
+/*
+    enum class WindowEvents
+    {
+        WindowClosedEvent, WindowResizedEvent, WindowFocusedEvent, WindowMovedEvent
+    };
+
+    // Events will be added here as they are needed.
+*/
```
window resized will be needed ASAP

colton-smith commented on Aquarius/Src/Aquarius/Events/MouseEvent.h:
```diff
@@ -0,0 +1,135 @@
+        }
+
+    private:
+        int m_inWindow;
```
Should probably be a bool

liamjwarren replied:
It has to be an int due to the GLFW callback, Daniel mentioned this before and thought it was something to do with C. That being said I have updated the inWindow() method to return a bool value rather than an int which could be used to check if the cursor is in the window. I will have this updated in my commit this evening.

colton-smith replied:
I mean you could cast the int returned by the callback to a bool before creating this event right? Doesn't really matter though

```
        static void cursorInWindowCallback(GLFWwindow* window, int inWindow)
        {
            bool bInWindow = inWindow;
            if (bInWindow)
            {
                AQ_INFO("Cursor has entered application window.");
                MouseEvent e(bInWindow)
            }

            else
            {
                AQ_INFO("Cursor has left application window.");
            }
        }
```

Its implicitly converted as well

liamjwarren replied:
Nvm, did a cast.

colton-smith commented on Aquarius/Src/Aquarius/Events/WindowEvent.h:
```diff
@@ -0,0 +1,17 @@
+
+#include "Event.h"
+
+#include <GLFW/glfw3.h>
+
```
Probably wont need this include

colton-smith commented on Aquarius/Src/Aquarius/Events/MouseEvent.h:
```diff
@@ -0,0 +1,135 @@
+
+#include "Event.h"
+
+#include <GLFW/glfw3.h>
+
```
Dont think this is used

colton-smith reviewed (changes requested):
Lots of small stuff to be changed, lmk if you have any questions

Burke-Daniel commented on Aquarius/Src/Aquarius/Core/Application.cpp:
```diff
@@ -33,6 +34,11 @@ namespace Aquarius {
 
+	void Application::onEvent(Event &event)
+	{
+	    EventHandler handler(event);
```
So it seems like any time onEvent is called from some piece of code (glfw callback functions for example), a new EventHandler instance is created, here (i.e constructor is called) and then it will immediately go out of scope. Do we potentially want to make EventHandler a member of the Application class, and that way then pieces of code that want to subscribe to certain events can access the handler through the static Application instance. Or I guess the alternative here is that we let the client handler the event stuff in their derived application class (though I assume we are going to want to have an EventHandler to reference through the glfw callback functions

colton-smith replied:
I didnt even see this change, I thought the Application class would just have one event handler instance

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/Callbacks.h:
```diff
@@ -0,0 +1,81 @@
+    static void keyboardCallback(GLFWwindow *window, int key, int scanCode, int action, int modifications)
+        {
+            switch (action) {
+                case GLFW_PRESS:
+                {
+                    KeyPressedEvent event(key);
+                    event.logEvent();
+                    break;
+                }
+
+                case GLFW_RELEASE:
+                {
+                    KeyReleasedEvent event(key);
+                    event.logEvent();
+                    break;
+                }
+
+                case GLFW_REPEAT:
+                {
+                    KeyRepeatedEvent event(key,1);
+                    event.logEvent();
+                    break;
+                }
```
So here for example, are we gonna want to have something like:
```C++
KeyPressedEvent event(key);
event.logEvent();
Application::get()->onEvent(event);
break;
```
where the call to Application::onEvent would then call it's event handler's `notify` method?

colton-smith replied:
Thats what I picture

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/EventHandler.h:
```diff
@@ -0,0 +1,55 @@
+
+    private:
+        std::unordered_map<eventType, std::vector<functionType>> subscribers;
+        Event& m_event;
```
Does EventHandler need to have reference to a particular Event? Cause notify is having an event passed into it. Only place this is used is on construction, though I suppose maybe derived classes will use it or something. To avoid having to pass in an Event to construct the Handler, what if m_event was instead an Event*?

colton-smith replied:
Why does the handler need an event at all? With this architecture, we shouldn't need that idt

Burke-Daniel replied:
Yeah that's why I was a bit confused about which direction we were going with the architecture of this

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/KeyboardEvent.h:
```diff
@@ -0,0 +1,67 @@
+    protected:
+        int m_keyCode;
+
+        KeyEvent(int keyCode) : m_keyCode(keyCode) {}
```
So if you use the Input::KeyCodes instead of ints here, you can just `static_cast<Input::KeyCode>(keyCode)` here and that should work

colton-smith replied:
To convert glfw to AQ keycodes?

Burke-Daniel replied:
Yeah

Burke-Daniel reviewed (changes requested):
Ok I can see this coming together better now, couple questions and couple comments for clarity, but main thing I am wondering is if you have decided to follow the observer pattern more tightly where the EventHandler object distributes the events to different places through the functions in the map, or more localized where there is a chain of event handling

liamjwarren committed:
Implementing review comments

Alright, here is my latest commit for this PR. A few things to note that are now implemented:
- logEvent() is now a virtual method
- Codes for Keyboard & Mouse events are now from the enums in Input.h
- Window Resize event and callback
- EventHandler updated based on comments
- Callbacks are in a Callbacks namespace
- Application has one instance of EventHandler which onEvent(Event &event) will not invoke the notify method of the event handler
- Callbacks are tied in to trigger onEvent now

liamjwarren commented:
Alright, here is my latest commit for this PR. A few things to note that are now implemented:
- logEvent() is now a virtual method
- Codes for Keyboard & Mouse events are now from the enums in Input.h
- Window Resize event and callback
- EventHandler updated based on comments
- Callbacks are in a Callbacks namespace
- Application has one instance of EventHandler which onEvent(Event &event) will not invoke the notify method of the event handler
- Callbacks are tied in to trigger onEvent now  

Let me know what you guys think!

liamjwarren committed:
Remove unnecessary includes in Application.cpp

liamjwarren committed:
Change m_RepeatCount to private

liamjwarren committed:
Add getter method for m_repeatCount and fixed offset

liamjwarren committed:
Fix the other offset

colton-smith commented on Aquarius/Src/Aquarius/Core/Window.h:
```diff
@@ -2,6 +2,7 @@
 
 #include "Aquarius/Core/Utility.h"
 #include "Aquarius/Renderer/RenderingContext.h"
+#include "Aquarius/Events/Event.h"
```
I dont see where this include is used

liamjwarren replied:
Oh Liam...

colton-smith commented on Aquarius/Src/Aquarius/Events/MouseEvent.h:
```diff
@@ -0,0 +1,133 @@
+            AQ_CORE_TRACE("MouseMovedEvent: %v, %v", m_XPosition, m_YPosition);
+        }
+
+        float getXCoordinate()
```
Maybe match the field name and getter name?

colton-smith reviewed (changes requested):
Super close to GTM:
- Need to update the getters and log methods to be const
- Need to break when the event is handled to avoid unnecessarily iteration
- Subscribers field in event handler should be prefixed m_ 

Then you should be g2g

liamjwarren committed:
Edits from review comments

- Const for log and getter methods
- break in EventHandler notify method
- Prefix the subscribers field of EventHandler

colton-smith commented on Aquarius/Aquarius.h:
```diff
@@ -16,9 +16,12 @@
-#include "Aquarius/Events/Keyboard.h"
-#include "Aquarius/Events/Mouse.h"
-#include "Aquarius/Events/Window.h"
+#include "Aquarius/Events/Callbacks.h"
```
This may be an exception - I doubt this needs to be exposed to the public

colton-smith reviewed (approved):
I added one note, not sure we need to have Callbacks.h in Aquarius.h as no one will need to use those except the Window class (everyone else can then sub to events).

GTM after @Burke-Daniel has a comb through, nice work!

liamjwarren committed:
Remove callback include in Aquarius.h

Burke-Daniel commented on Aquarius/Src/Aquarius/Events/KeyboardEvent.h:
```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#include "Event.h"
+
```
So I just wanted to test functionality of this, and it works but I had to add:
```C++
#include "Aquarius/Core/Input.h"
#include "Aquarius/Core/Log.h"
```
here.  But what I did to test was change using the escape key to close the window to be the application subscribing to a keyPressedEvent instead of using input polling. This is how I did that:
```C++
Application::Application(std::string&& windowName)
		: m_Window(Window::Create(800, 600, std::move(windowName)))
	{
		s_Application = this;

		Log::initLoggers();
		m_Window->Initialize();

		m_EventHandler.subscribe(eventType::KeyPressedEvent,
			[&](const Event& event)
			{
				try
				{
					const auto& keyEvent = dynamic_cast<const Aquarius::KeyPressedEvent&>(event);
					if (keyEvent.getCode() == Aquarius::Input::KeyCode::Key_escape)
					{
						glfwSetWindowShouldClose(m_Window->get(), true);
					}
				}
				catch (std::bad_cast&)
				{
					// Wrong event type, don't need to handle it
				}
			});

		AQ_CORE_INFO("Window Initialized Successfully");
	}
```
wIth a cheeky little lambda, and dynamic cast to cast the base Event reference down to a KeyPressedEvent so that I could call `getCode()`, this works to close the window as well, which is pretty sick!

Burke-Daniel reviewed (approved):
This is great! See my one comment about a couple of includes in KeyboardEvent.h and then GTM

liamjwarren committed:
Update includes in KeyboardEvent and MouseEvent

liamjwarren committed:
Merge branch 'master' into refine-initial-event-system

liamjwarren committed:
Fix minor formatting error
